File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Event/EventEmitter.swift
import Foundation

/**
 * EventEmitter
 *
 * Manages H5 event subscriptions (one handler per topic, overwrite on second on())
 * and pushes native events into the WebView.
 *
 * Subscriptions are keyed by topic. Emit dispatches a JSON payload through the
 * configured [JSExecutor], identical to the bridge response path.
 *
 * Thread safety: instances are @MainActor-isolated to match the rest of the
 * SDK; emit may be called from any actor / thread and dispatch hops to main.
 */
@MainActor
public final class EventEmitter {

    private let tag = "EventEmitter"
    private var topics: Set<String> = []

    /**
     * JS execution bridge. Wired by the host view controller to the WebView.
     * Nil before the WebView is ready (e.g. in unit tests).
     */
    public var jsExecutor: JSExecutor?

    /**
     * Register a handler for `topic`. Overwrites a previous handler for the
     * same topic (one handler per topic).
     */
    public func on(topic: String) {
        guard !topic.isEmpty else {
            Logger.shared.w(tag, "on rejected: empty topic")
            return
        }
        let wasPresent = topics.contains(topic)
        topics.insert(topic)
        Logger.shared.d(tag, "On: \(topic) (total=\(topics.count)\(wasPresent ? ", replaced previous" : ""))")
    }

    /**
     * Remove the handler for `topic`. No-op if not present.
     */
    public func off(topic: String) {
        if topics.remove(topic) != nil {
            Logger.shared.d(tag, "Off: \(topic) (total=\(topics.count))")
        }
    }

    /**
     * Broadcast a `topic` event to its registered handler (if any).
     *
     * The handler receives a JSON payload of shape `{topic, data}` and is
     * dispatched via `window.__coconutEvent('<json>')`. If no jsExecutor is
     * wired, the emit is silently dropped.
     */
    public func emit(topic: String, data: Any? = nil) {
        guard !topic.isEmpty else {
            Logger.shared.w(tag, "emit rejected: empty topic")
            return
        }
        guard topics.contains(topic) else {
            Logger.shared.d(tag, "emit no subscriber: \(topic)")
            return
        }
        dispatch(topic: topic, data: data)
    }

    /**
     * Emit bypassing the native subscription gate. Used for `nav.result`
     * delivery on container resume: the closing child's page load may have
     * clearAll()'d our native registration, but this page's own JS handler
     * table (coconut.on) is intact.
     */
    public func emitBypassingSubscription(topic: String, data: Any? = nil) {
        guard !topic.isEmpty else {
            Logger.shared.w(tag, "emit rejected: empty topic")
            return
        }
        dispatch(topic: topic, data: data)
    }

    private func dispatch(topic: String, data: Any?) {
        guard let executor = jsExecutor else {
            Logger.shared.w(tag, "emit dropped (no jsExecutor): \(topic)")
            return
        }

        let payload: [String: Any] = [
            "topic": topic,
            "data": data ?? NSNull(),
        ]

        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else {
            Logger.shared.e(tag, "Failed to encode payload for \(topic)")
            return
        }

        let escaped = Self.escapeJSString(json)
        let script = "window.__coconutEvent('\(escaped)')"

        Task { [weak self] in
            let error = await executor.evaluateJavaScript(script)
            if let error = error {
                Logger.shared.e(self?.tag ?? "EventEmitter",
                                "Failed to dispatch event for \(topic)",
                                error)
            }
        }
        Logger.shared.d(tag, "Emitted '\(topic)' (bypass=\(topics.contains(topic) ? "no" : "yes"))")
    }

    /**
     * Clear all subscriptions. Called on page navigation (reload / new URL)
     * to prevent stale subscribers from receiving events in a fresh page context.
     */
    public func clearAll() {
        let n = topics.count
        topics.removeAll()
        Logger.shared.d(tag, "Cleared \(n) subscription(s)")
    }

    /// Active subscription count (testing / diagnostics).
    public var count: Int { topics.count }

    /// Whether `topic` is currently subscribed (testing / diagnostics).
    public func has(topic: String) -> Bool { topics.contains(topic) }

    /**
     * Escape a raw JSON string for safe embedding inside a single-quoted
     * JavaScript string literal. Mirrors `BridgeResponseSender.sendResponse`.
     */
    static func escapeJSString(_ raw: String) -> String {
        return raw
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
    }
}

File: AndroidWebBox/coconut-core/src/main/java/com/sniper/coconut/event/EventEmitter.kt
package com.sniper.coconut.event

import com.sniper.coconut.utils.Logger
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.util.concurrent.ConcurrentHashMap

/**
 * EventEmitter
 *
 * Manages H5 event subscriptions (one handler per topic, overwrite on second on)
 * and pushes native events to the WebView.
 *
 * Subscriptions are keyed by topic. Emit dispatches a JSON payload via the
 * registered [jsExecutor], which the host (Activity / ViewController) wires to
 * evaluateJavascript on the UI thread.
 *
 * Threading: emit() may be invoked from any thread; the jsExecutor callback
 * is responsible for hopping to the WebView thread.
 */
class EventEmitter {

    /**
     * Subscribed topics. One handler per topic (overwrite semantics).
     */
    private val topics = ConcurrentHashMap.newKeySet<String>()

    /**
     * JS execution bridge. Set by the host when the WebView is ready.
     * Receives a complete JS statement, e.g. `window.__coconutEvent('...')`.
     */
    @Volatile
    var jsExecutor: ((script: String) -> Unit)? = null

    /**
     * Register a handler for [topic]. If already subscribed, the previous
     * registration is replaced (no-op in terms of count).
     */
    fun on(topic: String) {
        if (topic.isEmpty()) {
            Logger.w(TAG, "on rejected: empty topic")
            return
        }
        val wasPresent = !topics.add(topic)
        Logger.d(TAG, "On: $topic (total=${topics.size}${if (wasPresent) ", replaced previous" else ""})")
    }

    /**
     * Remove the handler for [topic]. No-op if not present.
     */
    fun off(topic: String) {
        if (topics.remove(topic)) {
            Logger.d(TAG, "Off: $topic (total=${topics.size})")
        }
    }

    /**
     * Broadcast a [topic] event to its registered handler (if any).
     *
     * Builds the wire payload `{topic, data}` and dispatches it through
     * [jsExecutor]. If no jsExecutor is wired, the emit is silently dropped
     * (e.g. during unit tests or before WebView is ready).
     */
    fun emit(topic: String, data: JsonElement? = null) {
        if (topic.isEmpty()) {
            Logger.w(TAG, "emit rejected: empty topic")
            return
        }
        if (!topics.contains(topic)) {
            Logger.d(TAG, "emit no subscriber: $topic")
            return
        }
        dispatch(topic, data)
    }

    /**
     * Dispatch [topic] to the page regardless of native subscription state.
     *
     * Used by nav.result draining: a backgrounded container's native
     * subscription may have been wiped by clearAll() when another container
     * loaded a page, but its JS handler table is intact (the page itself
     * never reloaded), so the wire dispatch still delivers.
     */
    fun emitBypassingSubscription(topic: String, data: JsonElement? = null) {
        if (topic.isEmpty()) {
            Logger.w(TAG, "emit rejected: empty topic")
            return
        }
        dispatch(topic, data)
    }

    private fun dispatch(topic: String, data: JsonElement?) {
        val executor = jsExecutor
        if (executor == null) {
            Logger.w(TAG, "emit dropped (no jsExecutor): $topic")
            return
        }

        val payload = buildJsonObject {
            put("topic", topic)
            put("data", data ?: JsonPrimitive(null as String?))
        }
        val json = json.encodeToString(JsonObject.serializer(), payload)
        val script = "window.__coconutEvent(${jsStringLiteral(json)})"
        try {
            executor.invoke(script)
        } catch (t: Throwable) {
            Logger.e(TAG, "Failed to dispatch event for $topic", t)
        }
        Logger.d(TAG, "Emitted '$topic'")
    }

    /**
     * Clear all subscriptions. Called on page navigation (reload / new URL)
     * to prevent stale subscribers from receiving events in a fresh page context.
     */
    fun clearAll() {
        val n = topics.size
        topics.clear()
        Logger.d(TAG, "Cleared $n subscription(s)")
    }

    /**
     * Number of active subscriptions (testing / diagnostics).
     */
    fun size(): Int = topics.size

    /**
     * Whether [topic] is currently subscribed (testing / diagnostics).
     */
    fun has(topic: String): Boolean = topics.contains(topic)

    /**
     * Encode a JSON string as a JavaScript string literal.
     *
     * The payload is itself JSON, so we wrap it in single quotes and escape
     * the few characters that can break out of a single-quoted JS string.
     * Forward slash is intentionally NOT escaped (optional in JSON spec).
     */
    private fun jsStringLiteral(raw: String): String {
        val sb = StringBuilder(raw.length + 8)
        sb.append('\'')
        for (c in raw) {
            when (c) {
                '\\' -> sb.append("\\\\")
                '\'' -> sb.append("\\'")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                else -> sb.append(c)
            }
        }
        sb.append('\'')
        return sb.toString()
    }

    companion object {
        private const val TAG = "EventEmitter"

        /**
         * Shared JSON encoder. Configured to be lenient on encoding policies
         * since payloads are already well-formed JsonObject instances.
         */
        @OptIn(ExperimentalSerializationApi::class)
        internal val json: Json = Json {
            encodeDefaults = true
            explicitNulls = false
        }
    }
}
